src/thor/centroid.cc

```diff
@@ -0,0 +1,248 @@
+  return location;
+}
+
+} // namespace
+
+namespace valhalla {
+namespace thor {
+
+// constructor
+PathIntersection::PathIntersection(uint64_t edge_id, uint64_t opp_id, uint8_t location_count)
+    : edge_id_(edge_id), opp_id_(opp_id) {
+  assert(location_count < 128);
+  // the smaller edge goes first for determinisms sake
+  if (opp_id_ < edge_id_)
+    std::swap(edge_id_, opp_id_);
+  // we pre-flip the bits of the paths we arent tracking so its as if they are already done
+  // that way its easy to tell when we are done later on
+  if (location_count < 64) {
+    lower_mask_ = ~((1ull << static_cast<uint64_t>(location_count)) - 1ull);
+    upper_mask_ = 0xffffffffffffffff;
+  } else {
+    lower_mask_ = 0;
+    upper_mask_ = ~((1ull << static_cast<uint64_t>(location_count - 64)) - 1ull);
+  }
+}
+
+// add an path has having connected at this intersection
+bool PathIntersection::AddPath(uint8_t path_id) const {
+  assert(path_id < 128);
+  if (path_id < 64) {
+    lower_mask_ |= 1 << static_cast<uint64_t>(path_id);
+  } else {
+    upper_mask_ |= 1 << static_cast<uint64_t>(path_id - 64);
+  }
+  // this will only be true once all the bits are flipped to true
+  return (lower_mask_ & upper_mask_) == 0xffffffffffffffff;
+}
+
+// has this path converged on this intersection (edge pair)
+bool PathIntersection::HasConverged(uint8_t path_id) const {
+  assert(path_id < 128);
+  if (path_id < 64) {
+    return lower_mask_ & (1 << static_cast<uint64_t>(path_id));
+  } else {
+    return upper_mask_ & (1 << static_cast<uint64_t>(path_id - 64));
+  }
+}
+
+// we only care about the edge id because that is what we hash on
+bool PathIntersection::operator==(const PathIntersection& i) const {
+  return edge_id_ == i.edge_id_;
+}
+
+// main entry point to the functionality
+std::vector<std::vector<PathInfo>>
+Centroid::forward(google::protobuf::RepeatedPtrField<valhalla::Location>& locations,
+                  baldr::GraphReader& reader,
+                  const sif::mode_costing_t& costings,
+                  const sif::TravelMode mode,
+                  valhalla::Location& centroid) {
+  // preflight check
+  if (locations.size() > baldr::kMaxMultiPathId)
+    throw std::runtime_error("Max number of locations exceeded");
+
+  // initialize state
+  location_count_ = locations.size();
+  best_intersection_ =
+      PathIntersection{baldr::kInvalidGraphId, baldr::kInvalidGraphId, location_count_};
+
+  // tell dijkstras we want to track the origin paths separately/concurrently
+  multipath_ = true;
+
+  // compute the expansion
+  Dijkstras::Compute(locations, reader, costings, mode);
+
+  // create the paths from the labelset
+  return FormPaths(locations, bdedgelabels_, reader, centroid);
+}
+
+// this is fired when the edge in the label has been settled (shortest path found) so we need to check
+// our intersections and add or update them
+thor::ExpansionRecommendation Centroid::ShouldExpand(baldr::GraphReader& reader,
```

Comment:
this is the meat of the algorithm. here we get informed that a specific location settled a specific edge, we check if we are already tracking it and we flip the bit corresponding to the input location that settled it. if we flipped the last outstanding bit in the mask, then its over and we found a least cost convergence point